### ISSUE TITLE
Should the structure be similar to the notebook spec? E.g., rendered markdown cells have no output

### ISSUE THREAD

**Issue by mpacer** [rgbkrk/next-gen-nteract-components#3]
I'm not sure about how I feel about the abstraction to put the rendered contents of Markdown cells in Output components. That definitely isn't how they're stored in nbformat… 

I feel like as a first pass keeping with the existing abstractions offered by the nbformat for the on disk representation would be a good aim?

I'm sure there was a good reason for doing it this way though. I'm interested to learn more!

**Comment by rgbkrk** [comment]
Thanks for asking!

The structure here is based on what the user experience is for both cells, which is that rendered markdown should line up vertically with output and code so that a user's eyes have a (hidden) vertical line to follow content.

![dashy](https://user-images.githubusercontent.com/836375/32672156-db70b5c8-c5fe-11e7-8cdb-365e79b9e471.png)

While the literal notebook format doesn't store rendered markdown as an output (and explicitly no `outputs` field in the notebook), the designation of being in the `<Outputs />` adheres to what nteract's current cells do which is to display an editor followed by the rendered view. The Editor is the input and the rendered view is the output. It's agnostic of the document format (which we also handle much differently in-memory).

It's less about being the _literal_ `<Outputs />` and more about being a part of the cell (UI-wise) that can be targeted, so that we can display other information inline where users see it immediately. This is mostly to make it easier for us to mock up new things like putting in information warning about IOPub limiting being in effect for a given cell (which would probably say that outputs are throttled rather than saying anything about IOPub):

```jsx
<Cell>
  <Input>
    <Prompt />
    <Editor />
  </Input>
  <Outputs>
    <WarningAboutIOPubLimiting />
    <SomeData />
  </Outputs>
</Cell>
```

As for Markdown cells, do they literally have to put rendered markdown in `<Outputs />`? No, I just wanted consistent padding while mocking up the notebooks. The rendered markdown area only has to contain `padding: 10px 10px 10px calc(var(--prompt-width, 50px) + 10px);` as well as a rule for when it's empty to `display: none` (or just be a `null` element) to match up well inside the cell.


```jsx
  <MarkdownCell>
    <Editor>{markdownSource}</Editor>
    <div className="buffered">
      <RenderedMarkdown>{markdownSource}</RenderedMarkdown>
      <style jsx>{`
        .buffered > :global(*) { /* <RenderedMarkdown /> could have this style or we'd have a container element */
          padding: 10px 10px 10px calc(var(--prompt-width, 50px) + 10px);
        }`}</style>
    </div>
  </MarkdownCell>
```

**Comment by rgbkrk** [comment]
This is funny -- I'm inconsistent with the original design of nteract, where the markdown doesn't line up and is off to the left:

![image](https://user-images.githubusercontent.com/836375/32674995-2a836362-c60a-11e7-805c-cd8a3e49b2e5.png)

I'm currently preferring the hidden vertical lines that user content flows down through.

**Comment by mpacer** [comment]
I think I agree that it's generally better to be aligned. I just am not sure that enforcing that via the document structure is the right way to go — it seems to blend styling and semantics a bit more than I like.

**Comment by rgbkrk** [comment]
What would be a good name for the container component that keeps the styling consistent between these? `<CellBottom />`?

**Comment by mpacer** [comment]
I guess I was thinking it shouldn't be a component but rather a style applied via css.

**Comment by rgbkrk** [comment]
In order to apply that style, you're going to have to target a selector of some sort. In this case it's a `className`, otherwise it's style you repeat in two different components and hope to keep consistent.

**Comment by mpacer** [comment]
Yes that sounds like exactly what I was intending to say. Using a className to apply a consistent style across different components via reference to that class & an appropriate css selector rule. I just don't think that we need to have to use the same `Component` class in order to apply a className consistently. 

The `RenderedMarkdown` component (as distinguished from the `Output` component associated with code cells) you use above is a perfectly reasonable approach. I could imagine wanting to target only `RenderedMarkdown` and not  `Output` and vice versa… for those things that should be shared (like the left alignment column) that should be accessed by reference to a className that they happen to share.

**Comment by rgbkrk** [comment]
What's the `className`? `cellBottom`?

**Comment by mpacer** [comment]
`cellLeftMargin`?

**Comment by rgbkrk** [comment]
Thanks, this helped me tease out what's the core here which is the notebook document margin. Semantically, with HTML (thinking these were web components):

## Code Cell

```
<prompt />  | <editor />
            | <outputs />
```

## Markdown Cell

### Edit mode

```
<prompt />  | <editor />
            | <renderedview />
```

### View mode

```
            | <renderedview />
```

**Comment by mpacer** [comment]
Should there be a prompt when you have a markdown cell?

Other than that, this is along the lines of what I was thinking.

**Comment by rgbkrk** [comment]
Oops I meant the prompt buffer which is a light indicator / grab handle.

**Comment by mpacer** [comment]
A grab handle should probably also exist in the render mode, at least on mouse hover.
